core/src/main/java/org/apache/accumulo/core/clientImpl/TableOperationsHelper.java

```diff
@@ -56,10 +57,12 @@ public void attachIterator(String tableName, IteratorSetting setting,
     for (IteratorScope scope : scopes) {
       String root = String.format("%s%s.%s", Property.TABLE_ITERATOR_PREFIX,
           scope.name().toLowerCase(), setting.getName());
-      for (Entry<String,String> prop : setting.getOptions().entrySet()) {
-        this.setProperty(tableName, root + ".opt." + prop.getKey(), prop.getValue());
-      }
-      this.setProperty(tableName, root, setting.getPriority() + "," + setting.getIteratorClass());
+
+      Map<String,String> propsToAdd = setting.getOptions().entrySet().stream()
+          .collect(Collectors.toMap(prop -> root + ".opt." + prop.getKey(), Entry::getValue));
+      propsToAdd.put(root, setting.getPriority() + "," + setting.getIteratorClass());
+
+      this.modifyProperties(tableName, props -> props.putAll(propsToAdd));
```

Comment:
ConcurrentModificationException is not being handled.  The previous code would set the props and never throw this.  Now a user could spuriously get this exception.    

I didn't realize that ConcurrentModificationException was a run time exception when looking at #2799.  I don't think that is good user experience for a spurious exception that user can not take any action to avoid and must handle.  Since its a run time exception when a developer is writing code they will most likely be unaware of it and it will most likely not be thrown in their testing.

Reply:
If no one else is working on it or opposed to it, I would like to try creating a PR for the new modifyProperties method to make it retry automatically instead of throwing the  ConcurrentModificationException.  If that was done, then this PR could stay as is and not have to handle that condition.

Reply:
> If no one else is working on it or opposed to it, I would like to try creating a PR for the new modifyProperties method to make it retry automatically instead of throwing the ConcurrentModificationException. If that was done, then this PR could stay as is and not have to handle that condition.

Sounds good to me :+1:

Reply:
Another option could be to catch and turn the ConcurrentModificationException into a CheckedException so that callers are forced to handle it.  If two (or more) separate processes are trying to set the same properties, a retry could effectively allow the last one to always "win".  If the catcher just wraps the CheckedException in a retry that's what would happen anyway - but it would be done explicitly. 

If the retry allows different, non-overlapping changes, that might be the "best" case.  P1 sets A, B, and P2 sets C, D then the final result should be A, B, C and D. The "worst" case would be if the final values were A,B or C,D 

If the sets are overlapping then the result would be non-determinant P1 sets A, deletes B and P2 sets B and C, then B could be set or deleted.  Changing the same property should be much less frequent than modifying non-overlapping sets.

Currently we do not guarantee that a process will see a given set of properties anyway - say for example a scan - it will execute with whatever it happens to find - if you set a property you need to allow it to propagate to the tservers, but something could change it before the scan actually starts and we do not specify otherwise.

Reply:
> Another option could be to catch and turn the ConcurrentModificationException into a CheckedException so that callers are forced to handle it. 

I am not a fan of this personally.  Looking at this PR, each call would have to handle the same case.  So that would lead to creating an internal static utility method that does the retry.

> If the retry allows different, non-overlapping changes, that might be the "best" case. P1 sets A, B, and P2 sets C, D then the final result should be A, B, C and D. The "worst" case would be if the final values were A,B or C,D
>
>If the sets are overlapping then the result would be non-determinant P1 sets A, deletes B and P2 sets B and C, then B could be set or deleted. Changing the same property should be much less frequent than modifying non-overlapping sets.

I created #2967 to do the retry and in that PR I also added a rigorous concurrency test.  The test makes lots of modifications in many threads and can detect if any single modification is lost.  The test also has some thread modify the same properties.

Ideally when multiple threads are concurrently modifying properties, that should result in the same outcome as if all those modifications were done serially in some order.  Barring bugs, conceptually I think the current accumulo code should be able to achieve this.